src/charm.py

```diff
@@ -659,25 +661,40 @@ def _on_start(self, event) -> None:
             event.defer()
             return
 
-    def _relation_changes_handler(self, event) -> None:
+    def _relation_changes_handler(self, event: RelationEvent) -> None:
         """Handles different relation events and updates MongoDB replica set."""
         self._connect_mongodb_exporter()
         self._connect_pbm_agent()
 
-        if type(event) is RelationDepartedEvent:
+        if isinstance(event, RelationDepartedEvent):
             if event.departing_unit.name == self.unit.name:
                 self.unit_peer_data.setdefault("unit_departed", "True")
 
         if not self.unit.is_leader():
             return
 
+        if self.upgrade_in_progress:
+            logger.warning(
+                "Adding replicas during an upgrade is not supported. The charm may be in a broken, unrecoverable state"
+            )
+            event.defer()
+            return
+
         # Admin password and keyFile should be created before running MongoDB.
         # This code runs on leader_elected event before mongod_pebble_ready
         self._generate_secrets()
 
         if not self.db_initialised:
             return
 
+        self._reconcile_mongo_data(event)
+
+    def _reconcile_mongo_data(self, event: RelationEvent) -> None:
```

Comment:
```suggestion
        self._reconcile_mongo_hosts(event)

    def _reconcile_mongo_hosts(self, event: RelationEvent) -> None:
```

Reply:
Thing is, it does not only reconcile mongo hosts but also users (as part of `update_app_relation_data`)

Reply:
hmm then maybe `_reconcile_mongo_hosts_and_users`? I think `data` is confusing here

`metadata` could also be good but I also think word is also vague

Reply:
Renamed to `_reconcile_mongo_hosts_and_users`